### PR TITLE
Add BSON depth tracking design

### PR DESCRIPTION
**Pull request by skriptble**

### CODE REVIEW

File: bson-depth-tracking/DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+validate methods augmented with additional depth tracking parameters to track how deep the
+validation should go. Each of these types has a `validate` method except for the `bson.Value` type.
+
+All of the internal calls during the validation process should use the `validate` method. This
+method will take the current depth and the maximum depth. An increase in the current depth number
+will occur before validating a BSON element of type `0x03`, `0x04`, or `0x0F`.  If, when entering
+the block for validation, the current depth is greater than the maximum depth, an
+`ErrMaximumDepthExceeded` error must be returned.
+
+To allow users to set the maximum depth, each of the public `Validate` methods will take an
```
Comment (craiggwilson) [inline bson-depth-tracking/DESIGN.md]: So, we are going to require users to provide this value? Most users won't care and won't want to think about it, and in almost all cases, don't need to.

**Comment by tychoish** [comment]
I don't think this is the right channel for this kind of content, but we are having that conversation offline. 

I think it'd be good if our designs could adopt the format the rest of the core engineering group uses for designs and scopes, at least from a high level and as needed, which will make it easier to review and consider various options. 

MongoDB doesn't accept bson that has more than 100 levels of nesting. [ref](https://docs.mongodb.com/manual/reference/limits/#Nested-Depth-for-BSON-Documents), so it seems like arbitrary configurable depth tracking might be over-engineered relative to the scope of the problem. If nothing else we should consider the pros and cons of 

I'd also be interested in seeing as part of this document, references to prior art in other drivers with regards to how they validate nested bson.

**Comment by skriptble** [comment]
This is a feature that was requested from users. I think some people may want to have a more restrictive nesting level. Additionally, the nesting level in MongoDB might change in the future, but that doesn't necessarily mean that users will also want the value to change for them.

Additionally, the work to do the tracking is the bulk of the problem and that work doesn't change if we make the depth static, it just means we potentially remove a function parameter. I would err on the side of flexibility in this case.

I'm not sure how other drivers do validation, but I'm not sure how applicable other drivers' BSON libraries will be to the Go driver's, in general.

File: bson-depth-tracking/DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+decode.
+
+## Scope of Work
+The `bson.Document`, `bson.Reader`, `bson.Element`, and `bson.Value` types need to have their
+validate methods augmented with additional depth tracking parameters to track how deep the
+validation should go. Each of these types has a `validate` method except for the `bson.Value` type.
+
+All of the internal calls during the validation process should use the `validate` method. This
+method will take the current depth and the maximum depth. An increase in the current depth number
+will occur before validating a BSON element of type `0x03`, `0x04`, or `0x0F`.  If, when entering
```
Comment (craiggwilson) [inline bson-depth-tracking/DESIGN.md]: Most people don't know what these are. 0x03 is a document, etc... might as well just spell it out.

File: bson-depth-tracking/DESIGN.md (hunk 1 above)
Comment (skriptble) [reply bson-depth-tracking/DESIGN.md]: We could have two methods, `Validate` and `ValidateWithDepth`, and we can have a default depth, perhaps 100 to match mongod's default?

**Comment by tychoish** [comment]
> > I'm not sure how other drivers do validation, but I'm not sure how applicable other drivers' BSON libraries will be to the Go driver's, in general.

I think it would be useful to potential implementors, and I would hope that the feature set of different drivers' are broadly equivalent. If this design isn't derived from prior art in other drivers, that feels important.

**Comment by skriptble** [comment]
> I think it would be useful to potential implementors, and I would hope that the feature set of different drivers' are broadly equivalent. If this design isn't derived from prior art in other drivers, that feels important.

The BSON library for the Go driver has a different design than the rest of the driver's BSON libraries, including mgo's BSON library. This is mostly because we had specific needs and needed a different performance profile than was provided by mgo's BSON library.

Additionally, to my knowledge, other drivers don't have types analogous to bson.Document nor to bson.Reader, so I don't think borrowing validation logic meant for differently designed types would be helpful here.

File: bson-depth-tracking/DESIGN.md (hunk 1 above)
Comment (craiggwilson) [reply bson-depth-tracking/DESIGN.md]: something like that, yes.

**Comment by tychoish** [comment]
> I don't think borrowing validation logic meant for differently designed types would be helpful here.

I can't imagine that there's any harm in doing this research as part of this design process and reflecting your findings here, to provide context in the future for this decisions.

**Comment by skriptble** [comment]
> I can't imagine that there's any harm in doing this research as part of this design process and reflecting your findings here, to provide context in the future for this decisions.

I think we can rely on Jeff and Craig for the context of what other drivers are doing and how this compares to that, so I don't think additional research is necessary here.

**Pull request closed without merging**